src/libstd/panic.rs

```diff
@@ -0,0 +1,194 @@
+/// advisory as it is not an unsafe trait. The `AssertPanicSafe` wrapper struct
+/// in this module can be used to force this trait to be implemented for any
+/// closed over variables passed to the `recover` function (more on this below).
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+#[rustc_on_unimplemented = "the type {Self} may not be safely transferred \
+                            across a recover boundary"]
+pub trait PanicSafe {}
+
+/// A marker trait representing types which do not contain an `UnsafeCell` by
+/// value internally.
+///
+/// This is a "helper marker trait" used to provide impl blocks for the
+/// `PanicSafe` trait, for more information see that documentation.
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+#[rustc_on_unimplemented = "the type {Self} contains interior mutability \
+                            and a reference may not be safely transferrable \
+                            across a recover boundary"]
+pub trait NoUnsafeCell {}
+
+/// A simple wrapper around a type to assert that it is panic safe.
+///
+/// When using `recover` it may be the case that some of the closed over
+/// variables are not panic safe. For example if `&mut T` is captured the
+/// compiler will generate a warning indicating that it is not panic safe. It
+/// may not be the case, however, that this is actually a problem due to the
+/// specific usage of `recover` if exception safety is specifically taken into
+/// account. This wrapper struct is useful for a quick and lightweight
+/// annotation that a variable is indeed panic safe.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(recover, std_panic)]
+///
+/// use std::panic::{self, AssertPanicSafe};
+///
+/// let mut variable = 4;
+///
+/// // This code will not compile becuause the closure captures `&mut variable`
+/// // which is not considered panic safe by default.
+///
+/// // panic::recover(|| {
+/// //     variable += 3;
+/// // });
+///
+/// // This, however, will compile due to the `AssertPanicSafe` wrapper
+/// let result = {
+///     let mut wrapper = AssertPanicSafe::new(&mut variable);
+///     panic::recover(move || {
+///         **wrapper += 3;
+///     })
+/// };
+/// // ...
+/// ```
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+pub struct AssertPanicSafe<T>(T);
+
+// Implementations of the `PanicSafe` trait:
+//
+// * By default everything is panic safe
+// * Mutable references are not panic safe
+// * &T where T contains interior mutability is not panic safe
+// * Types like Mutex/RwLock which are explicilty poisoned are panic safe
+// * Our custom AssertPanicSafe wrapper is indeed panic safe
+impl PanicSafe for .. {}
+impl<'a, T: ?Sized> !PanicSafe for &'a mut T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for &'a T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for *const T {}
+impl<'a, T: NoUnsafeCell + ?Sized> PanicSafe for *mut T {}
```

Comment:
Hm, I'm surprised that was missing from the RFC; it was a key part of the earlier discussion. It's important for the FFI case in particular, and since using the pointers requires unsafe code anyway the need for tailored assertions like `AssertPanicSafe` is lessened. 

Remember that this trait is unlike the other OIBITs, by design, in several respects -- this being one of them. It's also a safe trait to implement. Like the existing approach to mitigating panic safety issues, the trait acts primarily as a "speed bump" when isolation boundaries may be violated.